status/views.py

```diff
@@ -1,97 +1,109 @@
+            if rs.ping() == True:
+                status = 'online'
             else:
-                data[2]['status'] = 'online'
-        else:
-            data[2]['status'] = 'unknown'
-            data[2]['details'] = 'Error connecting to /status'
-    except Exception as e:
-        data[2]['status'] = 'offline'
-        data[2]['details'] = str(e)
-
-    if settings.SEND_MAIL:
-        data.append(
-        {
-            "name": "Mail Daemon",
-            "status": "unknown",
-            "details": ''
-        })
+                status = 'offline'
+                details = "PING failed"
+        except redis.exceptions.RedisError as e:
+            status = 'offline'
+            details = "PING failed:\n" + str(e)
+
+        return status, details
+
+    def test_postgresql_connection(self):
+        status, details = 'unknown', ''
+
         try:
-            response = requests.get(settings.MAIL_SOCK_URL)
-            if response.status_code != 200:
-                data[3]['status'] = 'unknown'
-                data[3]['details'] = 'Unknown error whilst connecting to /'
+            with connection.cursor() as cursor:
+                cursor.execute("SELECT 1");
+            status = 'online'
+        except psycopg2.OperationalError as e:
+            status = 'offline'
+            details = 'Could not execute query `SELECT 1\':\n' + str(e)
+        
+        return status, details
+
+    def get_andromeda_status(self):
+        status, details = 'unknown', ''
+        try:
+            response = requests.get(f"{settings.CHALLENGE_SERVER_URL}/status",
+                                headers={"Authorization": settings.CHALLENGE_SERVER_API_KEY})
+            if response.status_code == 200:
+                sysinfo = response.json()
+                reasons = []
+
+                if sysinfo['Loads'][0] > 90:
+                    reasons.append(f"High CPU load: {sysinfo['Loads'][0]}")
+
+                if sysinfo['FreeRam'] / sysinfo['TotalRam'] < 0.2:
+                    reasons.append(f"High RAM usage: {sysinfo['FreeRam'] / 1024}MB free out of {sysinfo['TotalRam']}")
+
+                if len(reasons) != 0:
+                    status = 'partial'
+                    details = '\n'.join(reasons)
+                else:
+                    status = 'online'
             else:
-                data[3]['status'] = 'online'
-        except Exception as e:
-            data[3]['status'] = 'offline'
-            data[3]['details'] = str(e)
+                status = 'unknown'
+                details = 'Error connecting to /status'
+        except requests.exceptions.RequestException as e:
+            status = 'offline'
+            details = str(e)
+        
+        return status, details
 
+    def test_mailusv_connection(self):
+        status, details = 'unknown', ''
 
-    return FormattedResponse(data)
+        try:
+            response = requests.get(settings.MAIL_SOCK_URL)
+            if response.status_code != 200:
```

Comment:
and similar to what I said in the andromeda comment, if this is returning something that isnt 200, something has gone seriously wrong and it should be assumed to be down